src/tree/updater_quantile_hist.cc

```diff
@@ -102,33 +104,82 @@ void QuantileHistMaker::Builder::SyncHistograms(
+      CopyHist(this_local, this_hist, r.begin(), r.end());
+
+      if (!(*p_tree)[entry.nid].IsRoot() && entry.sibling_nid > -1) {
+        auto sibling_hist = hist_[entry.sibling_nid];
+        auto parent_hist = phist_local_[(*p_tree)[entry.nid].Parent()];
+        SubtractionHist(sibling_hist, parent_hist, this_hist, r.begin(), r.end());
+        // Store posible parent node
+        auto sibling_local = phist_local_[entry.sibling_nid];
+        CopyHist(sibling_local, sibling_hist, r.begin(), r.end());
+      }
+    });
 
-  common::ParallelFor2d(space, this->nthread_, [&](size_t node, common::Range1d r) {
-    const auto entry = nodes_for_explicit_hist_build_[node];
-    auto this_hist = hist_[entry.nid];
-    // Merging histograms from each thread into once
-    hist_buffer_.ReduceHist(node, r.begin(), r.end());
+    builder_monitor_.Start("SyncHistogramsAllreduce");
+    this->histred_.Allreduce(hist_[starting_index].data(), hist_builder_.GetNumBins() * sync_count);
+    builder_monitor_.Stop("SyncHistogramsAllreduce");
 
-    if (!(*p_tree)[entry.nid].IsRoot() && entry.sibling_nid > -1 && !isDistributed) {
-      auto parent_hist = hist_[(*p_tree)[entry.nid].Parent()];
-      auto sibling_hist = hist_[entry.sibling_nid];
+    common::ParallelFor2d(space, this->nthread_, [&](size_t node, common::Range1d r) {
+      const auto entry = nodes_for_explicit_hist_build_[node];
+      if (!((*p_tree)[entry.nid].IsLeftChild())) {
+        auto this_hist = hist_[entry.nid];
 
-      SubtractionHist(sibling_hist, parent_hist, this_hist, r.begin(), r.end());
-    }
-  });
+        if (!(*p_tree)[entry.nid].IsRoot() && entry.sibling_nid > -1) {
+          auto parent_hist = hist_[(*p_tree)[entry.nid].Parent()];
+          auto sibling_hist = hist_[entry.sibling_nid];
+          SubtractionHist(this_hist, parent_hist, sibling_hist, r.begin(), r.end());
+        }
+      }
+    });
 
-  if (isDistributed) {
-    this->histred_.Allreduce(hist_[starting_index].data(), hist_builder_.GetNumBins() * sync_count);
-    // use Subtraction Trick
-    for (auto const& node : nodes_for_subtraction_trick_) {
-      SubtractionTrick(hist_[node.nid], hist_[node.sibling_nid],
-                       hist_[(*p_tree)[node.nid].Parent()]);
-    }
+    common::BlockedSpace2d space2(nodes_for_subtraction_trick_.size(), [&](size_t node) {
+      return nbins;
+    }, 1024);
+
+    common::ParallelFor2d(space2, this->nthread_, [&](size_t node, common::Range1d r) {
+      const auto entry = nodes_for_subtraction_trick_[node];
+      if (!((*p_tree)[entry.nid].IsLeftChild())) {
+        auto this_hist = hist_[entry.nid];
+
+        if (!(*p_tree)[entry.nid].IsRoot() && entry.sibling_nid > -1) {
+          auto parent_hist = hist_[(*p_tree)[entry.nid].Parent()];
+          auto sibling_hist = hist_[entry.sibling_nid];
+          SubtractionHist(this_hist, parent_hist, sibling_hist, r.begin(), r.end());
+        }
+      }
+    });
```

Comment:
So, you just do Subtraction Trick for left nodes from 2 different queues, other things are the same.
You can merge them into one simply or extract to a new method. Anyway, we should reduce "copy-past" code.

Reply:
method `ParallelSubtractionHist` was added